### ISSUE TITLE
Silent Notifications failing to call push.on('notification') in the background with iOS 11

### ISSUE THREAD

**Issue by DaDanny** [phonegap/phonegap-plugin-push#1977]
### Expected Behaviour
When a notification which includes`{
     "content-available": 1
}` in the `aps` object is sent to a device running iOS 11 with the app running in the background, the push message should signal iOS to wake up the app and begin background processing. 

### Actual Behaviour
On iOS 11, the push message no longer signals to iOS to wake up the app for background processing. The `push.on('notification')` method is only called if a user clicks on the notification.

### Reproduce Scenario (including but not limited to)
Initialize the plugin following the README example code:

```
var push = PushNotification.init({
	"ios": {
		"sound": "true",
		"alert": "true",
		"badge": "true",
		"clearBadge": "true"
	}
});

push.on('registration', function(data) {
	// send data.registrationId to push service
});


push.on('notification', function(data) {
	// do something with the push data
	// then call finish to let the OS know we are done

        // Additional log statement to help with debugging
        console.log('Notification received');

	push.finish(function() {
		console.log("processing of push data is finished");
	}, function() {
        console.log("something went wrong with push.finish for ID = " + data.additionalData.notId)
    }, data.additionalData.notId);
});
```

Then send a notification with the following JSON:
```
{
    "aps": {
        "alert": "Test background push",
        "content-available": 1
    },
    "notId": 1 // unique ID you generate
}
```

- If the app is running in the foreground, the notification behaves as expected and the `push.on('notification')` event handler is properly called
- If the app is running in the background on a device running iOS 11.x (GM or beta), the `push.on('notification')` event handler is not called until a user clicks on the notification.
- I've noticed that when I start the app via XCode, the event handler will be properly called in the background, but once I stop the app in XCode and launch it via my device, the event handler is no longer called in the background. 

#### Steps to Reproduce 
- **Device must be running iOS 11**
- Initialize the plugin with the code above or with the example code provided in the README file. 
- Launch the app with XCode
- Send the app to the background by switching to another app or clicking the home button
- Send a notification with `{
     "content-available": 1
}` in the `aps` object. 
- Verify that you see the notification log statements (in XCode or Safari inspector)
- Stop running the app in XCode (click the stop button)
- Completely close out of the app on your device
- Launch the app via your device (how your users launch the app)
- Open up Safari to monitor your app's log statements.
- Send the app to the background again
- Send the same push notification with `content-available : 1`
- The `push.on('notification')` event handler will not be called while the app is in the background. If you click the notification, the app will return to the foreground and the handler will be called.

#### Platform and Version (eg. Android 5.0 or iOS 9.2.1)

iOS : 11.0 (GM)

#### (Android) What device vendor (e.g. Samsung, HTC, Sony...)

N/A

#### Cordova CLI version and cordova platform version

    cordova --version                                    7.0.1
    cordova platform version ios                   4.5.0-dev

#### Plugin version

    cordova plugin version | grep phonegap-plugin-push   phonegap-plugin-push 2.0.0 "PushPlugin"

#### Sample Push Data Payload

```
{
    "aps": {
        "alert": "Test background push",
        "content-available": 1
    },
    "notId": 1 // unique ID you generate
}
```

#### Sample Code that illustrates the problem

Code above

#### Logs taken while reproducing problem

Looking at my device logs via the macOS Console app, I see a few that seem to be the root of the issue:

- `Received remote notification request 5F54-336B [ hasAlertContent: 1, hasSound: 1 hasBadge: 0 hasContentAvailable: 1 hasMutableContent: 0 ] `

- `Deliver push notification request 5F54-336B`

- `Passing content-available push to Duet`

- `Removing a launch request for application <private> by activity <private>`

#### Additional Details

It seems as though the notifications will behave as expected when the app is launched via XCode, but the notification event handler is not being called when the app is running in the background after being launched without XCode. 

I found this stackoverflow post where other people are experiencing this issue as well:

[Stackoverflow](https://stackoverflow.com/questions/44796613/silent-pushes-not-delivered-to-the-app-on-ios-11)

which leads me to believe that this is the intended behavior on iOS 11, but I'm hoping someone here has experienced this too and hopefully knows of a solution. 

Thanks.

**Comment by aleppos** [comment]
I am having similar issue,

IOS push.on('notification') not triggered when app is closed.

**Comment by coolvasanth** [comment]
Have you guys tried keeping the code inside platform ready function? (if your using Ionic then you should enclose your push notification code inside 
`this.platform.ready().then( e => {}`
)

**Comment by DaDanny** [comment]
Hey @coolvasanth ,
Thanks for the tip. I just double checked and my code is already wrapped within a platform ready function. I think I would be seeing other errors if it wasn't.

Also, when I launch the app via XCode, everything behaves correctly, but once I stop the app and force close it, then launch it again via my device, everything still works correctly except for the background event handler being called. 

That is what makes this somewhat tough, as when you are testing on your device, its very likely that you are launching the app via XCode and testing it that way, which works. But once you disconnect or stop the app then relaunch it via your device, it no longer behaves as expected.


Unfortunately, I don't believe this is a bug with this library, but with iOS 11, though I'd still like to know if other's have experienced this and if they have any solutions.

**Comment by aleppos** [comment]
The same code working fine on Android and on IOS only when app is running.

**Comment by DaDanny** [comment]
Hi @macdonst , first, thanks for this awesome plugin, it is truly incredible.

However, I'm not sure what the "retest" label is meant for here. Are you saying I should retest my code and make sure I'm following all the steps correctly? 

When I posted this issue, one thing I was hoping to accomplish is discovering if anyone else has experienced this issue with iOS 11, so I can start ruling out my implementation of the plugin if iOS 11 behaves differently.

**Comment by mattcollins171** [comment]
Yip having the same issue, nothing happens when the app is in the background, but when it's loaded to the foreground, the functionality that I want to happen in the background kicks in.

**Comment by fredgalvao** [comment]
@DaDanny `retest` tag means **we** need to retest it to make sure it's not broken. Considering iOS 11 is quite new, I'd say give it a few days and someone will come up with a pull request to fix it, if it's indeed broken. Or you can help us and provide the PR yourself ;)

**Comment by DaDanny** [comment]
Hey @fredgalvao , thanks for clarifying. 

I would be more than happy to provide a PR provided I'm able to come up with a proper solution but first I wanted to see if other people were experiencing this issue before I assumed the error was of my own. 

I am currently looking into this issue and will keep this thread updated with anything I find. There is an active on-going discussion over on Stackoverflow in addition to the link I included in the first post, and I'm checking that daily for any new info.

[Stackoverflow Discussion](https://chat.stackoverflow.com/rooms/155203/discussion-between-jan-and-alexwoe89)

For now, I did manage to come up with a "band-aid" solution for my own application, but it is completely outside the scope of this library and maybe too specific to my app, though I'm happy to share it here if anyone is interested or really struggling with a workaround. 

Thanks again for this awesome plugin,

Cheers!

**Comment by DaDanny** [comment]
Good news, Apple just released the beta version of 11.1 and in the release notes they state

> Notifications Resolved Issues • Silent push notifications are processed more frequently. (33278611)

After updating my device and testing it using the steps I outlined in the OP, it looks like this is no longer and issue with iOS 11.1 and silent/background notifications are now working correctly.

**Comment by fredgalvao** [comment]
Good to know (that Apple is still good 'ol Apple)! Thanks @DaDanny 

@aleppos @mattcollins171 can you guys confirm that updating the device to 11.1 fixes this intermittency?

If so, we can close this (unless @macdonst wants to retest anyway).

**Comment by macdonst** [comment]
@fredgalvao seems like something we should mention in the docs.

**Comment by fredgalvao** [comment]
We'd need confirmation. I unfortunately don't have the resources to test it.

**Comment by mattcollins171** [comment]
I can confirm from my end that 11.1 now seems to be working.

**Comment by macdonst** [comment]
Okay, I've added a note in the docs and will close this for now.

**Comment by DaDanny** [comment]
Awesome, thanks for the fast response guys. I will keep testing and let you guys know if I am still experiencing any issues. 

Other than that, this is an incredible plugin and very well put together. Thanks again,

Cheers

**Comment by ElleDriver** [comment]
Hi DaDanny, you said (a long time ago) "For now, I did manage to come up with a "band-aid" solution for my own application, but it is completely outside the scope of this library and maybe too specific to my app, though I'm happy to share it here if anyone is interested or really struggling with a workaround." 
I am still having issues with iOS 11 and remote notifications sound (I am unable to start an alert sound when my app is in background and a notification is received). May be your workaround (outside the scope of this library) could help me? Thanks!

**Comment by DaDanny** [comment]
Hey @ElleDriver, 

First, I love your username, I'm a huge Kill Bill fan :)

Unfortunately though, I don't think my workaround will help you with your issue. Have you been able to get a notification to play an alert sound before? Does the device get any notification at all? 

Without more details on your issue, I think it may be related to the payload you are sending to the device, but it could be a few other things. 

If I were you, I would double check your code to make sure you are setting all the proper fields on the notification payload and if everything looks correct but still won't behave properly, open a new issue. 


As for my "band-aid" solution, I am building a chat application and use notifications when a user is in a chat. I wanted to keep the user's chats up-to-date so when they open the app or click a notification, they don't need to wait for the chat to refresh with the latest messages. 

With Android and the latest iOS 11 release, I can use background-notifications to store incoming messages so when the user returns to a chat, it is all up to date with the latest messages.

Prior to the iOS 11.1 fix, background-notifications weren't triggering the function to process notifications if the app was suspended, so when a user clicked a notification, I could only process one message at a time, which means if User A sends 5 messages to User B, User B may get 5 notifications, but only the last message would be available until the chat refreshed since typically one message = one notification.

I ended up having to do a few things to work around this issue and make sure the chat was up to date with all messages.

First, I grouped chat notifications with an ID unique to each chat and used this as the 'collapseId' on the notification payload. This meant that each chat would have only one notification, regardless of how many messages were sent out. So in the case with User A sending User B 5 messages, User B would only see a single notification for that chat. 

Then, before I sent out the notification, I created a new array field called "lastMessages" which I include with the notification payload. "lastMessages" is populated by iterating in reverse through the chat's messages, but before I push a message to "lastMessages", I double check the size of the payload in order to make sure I'm still under the maximum size for notifications, which I believe is 2kb. If adding a message keeps me under the payload size constraint, I push it to the array "lastMessages".

Once I reach my size constraint or run out of messages, I send the notification payload which has that "lastMessages" field included.

Then on the app side of things, when a user clicks a notification and we begin processing it, we can compare the current message array with the "lastMessages" array in order to make sure the chat has the latest messages. 


It has been working pretty well, but after the fix with iOS 11.1, it is a little overkill.

Anyways though, I hope that clears things up for you and maybe even helps you out. Let me know if you have any questions.


Danny
